collector/tcp.go

```diff
@@ -0,0 +1,249 @@
+			nil,
+			nil,
+		),
+		FrequencyPerfTime: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "frequency_perftime"),
+			"(TCP.FrequencyPerfTime)",
+			nil,
+			nil,
+		),
+		FrequencySys100NS: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "frequency_sys100ns"),
+			"(TCP.FrequencySys100NS)",
+			nil,
+			nil,
+		),
+		SegmentsPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "segments_per_sec"),
+			"(TCP.SegmentsPerSec)",
+			nil,
+			nil,
+		),
+		SegmentsReceivedPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "segments_received_per_sec"),
+			"(TCP.SegmentsReceivedPerSec)",
+			nil,
+			nil,
+		),
+		SegmentsRetransmittedPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "segments_retransmitted_per_sec"),
+			"(TCP.SegmentsRetransmittedPerSec)",
+			nil,
+			nil,
+		),
+		SegmentsSentPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "segments_sent_per_sec"),
+			"(TCP.SegmentsSentPerSec)",
+			nil,
+			nil,
+		),
+		TimestampObject: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "timestamp_object"),
+			"(TCP.TimestampObject)",
+			nil,
+			nil,
+		),
+		TimestampPerfTime: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "timestamp_perftime"),
+			"(TCP.TimestampPerfTime)",
+			nil,
+			nil,
+		),
+		TimestampSys100NS: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "timestamp_sys100ns"),
+			"(TCP.TimestampSys100NS)",
+			nil,
+			nil,
+		),
+	}, nil
+}
+
+// Collect sends the metric values for each metric
+// to the provided prometheus Metric channel.
+func (c *TCPCollector) Collect(ch chan<- prometheus.Metric) error {
+	if desc, err := c.collect(ch); err != nil {
+		log.Println("[ERROR] failed collecting tcp metrics:", desc, err)
+		return err
+	}
+	return nil
+}
+
+type Win32_PerfRawData_Tcpip_TCPv4 struct {
+	ConnectionFailures          uint64
+	ConnectionsActive           uint64
+	ConnectionsEstablished      uint64
+	ConnectionsPassive          uint64
+	ConnectionsReset            uint64
+	Frequency_Object            uint64
+	Frequency_PerfTime          uint64
+	Frequency_Sys100NS          uint64
+	SegmentsPersec              uint64
+	SegmentsReceivedPersec      uint64
+	SegmentsRetransmittedPersec uint64
+	SegmentsSentPersec          uint64
+	Timestamp_Object            uint64
+	Timestamp_PerfTime          uint64
+	Timestamp_Sys100NS          uint64
+}
+
+func (c *TCPCollector) collect(ch chan<- prometheus.Metric) (*prometheus.Desc, error) {
+	var dst []Win32_PerfRawData_Tcpip_TCPv4
+
+	q := wmi.CreateQuery(&dst, "")
+	if err := wmi.Query(q, &dst); err != nil {
+		return nil, err
+	}
+
+	// Counters
+	ch <- prometheus.MustNewConstMetric(
+		c.ConnectionFailures,
+		prometheus.CounterValue,
+		float64(dst[0].ConnectionFailures),
+	)
+	ch <- prometheus.MustNewConstMetric(
+		c.ConnectionsActive,
+		prometheus.CounterValue,
```

Comment:
Is this actually a counter? I would have expected a gauge?

Reply:
Strangely it is a counter. I graphed them overnight as a gauge, but that resulted in eternally climbing values. Or is that because I indicated it to be a counter in code?

Reply:
When reading https://msdn.microsoft.com/en-us/library/aa394341(v=vs.85).aspx the Connection*** values seem to be counters, but it's open for debate. I copied the use of CounterValue from the `net` collector.

Treating them as counters results in a graph like this, which seems about right:

![tcp-counters](https://user-images.githubusercontent.com/136472/34149185-915afcea-e4a4-11e7-9f42-4d401f4c834a.png)

Reply:
Huh... Oh well, it is not the first time that things have names that don't really match. Good that you checked!